File: server/src/langchain/callbacks/streaming-callback-handler.ts
import { BaseCallbackHandler } from 'langchain/callbacks';
import {
  ChainValues,
  AgentAction,
  AgentFinish,
  LLMResult
} from 'langchain/schema';

class StreamingCallbackHandler extends BaseCallbackHandler {
  name = 'StreamingCallbackHandler';
  private static streamCallback?: (token: string) => void;

  static setStreamCallback(callback: (token: string) => void) {
    StreamingCallbackHandler.streamCallback = callback;
  }

  async handleChainEnd(_output: ChainValues) {
    if (
      StreamingCallbackHandler.streamCallback &&
      process.env.CHAIN_END_TRIGGER_MESSAGE
    ) {
      StreamingCallbackHandler.streamCallback(
        process.env.CHAIN_END_TRIGGER_MESSAGE
      );
    }
  }

  async handleLLMNewToken(token: string) {
    StreamingCallbackHandler.streamCallback &&
      StreamingCallbackHandler.streamCallback(token);
  }
}

export { StreamingCallbackHandler };

File: pyServer/server/langchain/callbacks/streaming_callback_handler.py
import os
from dotenv import load_dotenv
from langchain.callbacks.base import BaseCallbackHandler
from langchain.schema import LLMResult
from typing import Dict, Any, List

load_dotenv()


class StreamingCallbackHandler(BaseCallbackHandler):
    _stream_callback = None

    @classmethod
    def set_stream_callback(cls, function):
        StreamingCallbackHandler._stream_callback = staticmethod(function)

    @property
    def stream_callback(self):
        return self._stream_callback

    @stream_callback.setter
    def stream_callback(self, function):
        StreamingCallbackHandler.set_stream_callback(function)

    def on_llm_new_token(self, token: str, **kwargs: Any) -> Any:
        if self.stream_callback:
            self.stream_callback(token)

    def on_llm_start(
        self, serialized: Dict[str, Any], prompts: List[str], **kwargs: Any
    ) -> Any:
        if self.stream_callback:
            self.stream_callback(os.getenv("LLM_START_TRIGGER_MESSAGE"))

    def on_chain_end(self, outputs: Dict[str, Any], **kwargs: Any) -> Any:
        if self.stream_callback:
            self.stream_callback(os.getenv("CHAIN_END_TRIGGER_MESSAGE"))
